File: SOURCE_NATIVE/EW4_iOS_NativePort/NativeCore/Sources/EW4NativeCore/NativeAchievementCore.swift
import Foundation

public struct NativeAchievementStatSummary: Equatable, Sendable {
    public let level: Int
    public let score: Int
    public let raw: Int
}

public struct NativeAchievementContinent: Equatable, Sendable {
    public let value: Int
    public let display: Int
    public let digits: [Int]
}

public struct NativeAchievementViewModel: Equatable, Sendable {
    public let stageStarsEarned: Int
    public let stageStarsMax: Int
    public let military: NativeAchievementStatSummary
    public let nobility: NativeAchievementStatSummary
    public let continents: [String: NativeAchievementContinent]
    public let generalIDs: [Int]
}

/// Swift parity for mature P28 `native_achievement_core.js` local semantics.
/// Champion/Ranking platform navigation remains intentionally outside the
/// single-player PORT_ONLY runtime.
public enum NativeAchievementCore {
    public static let stageCounts = [18, 17, 15, 13, 11, 10]
    public static let maxStageStars = stageCounts.reduce(0) { $0 + $1 * 5 }

    public static func ownedGeneralIDs(profile: NativePlayerProfile, maxID: Int = 208) -> [Int] {
        var seen = Set<Int>()
        return profile.ownedCommanderIDs.filter { id in
            guard id >= 1, id <= maxID, !seen.contains(id) else { return false }
            seen.insert(id); return true
        }
    }

    public static func rawMilitary(profile: NativePlayerProfile, commander: Commander) -> Int {
        let state = NativePlayerProfileCore.generalGrowthState(profile: profile, raw: commander)
        var total = 300 + max(0, state.militaryProgress)
        if state.rank > 0 {
            for i in 0..<min(state.rank, NativeGeneralGrowthCore.militaryThresholds.count) {
                total += NativeGeneralGrowthCore.militaryThresholds[i]
            }
        }
        return total
    }

    public static func rawNobility(profile: NativePlayerProfile, commander: Commander) -> Int {
        let state = NativePlayerProfileCore.generalGrowthState(profile: profile, raw: commander)
        var total = 60 + max(0, state.nobilityProgress)
        if state.nobility > 0 {
            for i in 0..<min(state.nobility, NativeGeneralGrowthCore.nobilityThresholds.count) {
                total += NativeGeneralGrowthCore.nobilityThresholds[i]
            }
        }
        return total
    }

    public static func aggregateLevel(total: Int, cutoff: Int, first: Int, multiplier: Float) -> Int {
        let total = max(0, total)
        if total <= cutoff { return 1 }
        var level = 2
        var threshold = first
        while level < 99 && total >= threshold {
            threshold = Int(Float(threshold) * multiplier)
            level += 1
        }
        return level
    }

    public static func globalSummary(profile: NativePlayerProfile, commanders: [Int: Commander]) -> (military: NativeAchievementStatSummary, nobility: NativeAchievementStatSummary) {
        let generals = ownedGeneralIDs(profile: profile).compactMap { commanders[$0] }
        let militaryRaw = generals.reduce(0) { $0 + rawMilitary(profile: profile, commander: $1) }
        let nobilityRaw = generals.reduce(0) { $0 + rawNobility(profile: profile, commander: $1) }
        return (
            .init(level: aggregateLevel(total: militaryRaw, cutoff: 99, first: 121, multiplier: 1.214), score: militaryRaw / 10, raw: militaryRaw),
            .init(level: aggregateLevel(total: nobilityRaw, cutoff: 49, first: 56, multiplier: 1.125), score: nobilityRaw * 4, raw: nobilityRaw)
        )
    }

    public static func campaignStageStars(profile: NativePlayerProfile) -> (earned: Int, max: Int) {
        let best = profile.intMap("campaignBestRating")
        let progress = profile.intMap("campaignProgress")
        var earned = 0
        for zone in 1...stageCounts.count {
            for stage in 1...stageCounts[zone - 1] {
                let file = "campaign\(zone)_\(String(format: "%02d", stage)).btl"
                if let rating = best[file] { earned += min(5, max(0, rating)) }
                else {
                    let p = min(2, max(0, progress[file] ?? 0))
                    earned += p >= 2 ? 5 : (p >= 1 ? 1 : 0)
                }
            }
        }
        return (min(maxStageStars, max(0, earned)), maxStageStars)
    }

    public static func continentStoredValue(profile: NativePlayerProfile, key: String) -> Int {
        guard case .object(let records) = profile.document["achievementConquests"], let raw = records[key] else { return 0 }
        let value: Int
        switch raw {
        case .int(let v): value = v
        case .double(let v): value = Int(v)
        case .string(let v): value = Int(v) ?? 0
        case .object(let object):
            if let item = object["value"] {
                switch item { case .int(let v): value=v; case .double(let v): value=Int(v); case .string(let v): value=Int(v) ?? 0; default: value=0 }
            } else if case .array(let digits)? = object["rule"] {
                value = digits.prefix(3).reduce(0) { partial, item in
                    let d: Int
                    switch item { case .int(let v): d=v; case .double(let v): d=Int(v); case .string(let v): d=Int(v) ?? 0; default: d=0 }
                    return partial * 10 + min(9, max(0, d))
                }
            } else { value = 0 }
        default: value = 0
        }
        return min(1000, max(0, value))
    }

    public static func ruleDigits(_ value: Int) -> [Int] {
        String(min(999, max(0, value))).compactMap { Int(String($0)) }
    }

    public static func continent(profile: NativePlayerProfile, key: String) -> NativeAchievementContinent {
        let stored = continentStoredValue(profile: profile, key: key)
        let display = min(999, stored)
        return .init(value: stored, display: display, digits: ruleDigits(display))
    }

    public static func viewModel(profile: NativePlayerProfile, commanders: [Int: Commander]) -> NativeAchievementViewModel {
        let stars = campaignStageStars(profile: profile)
        let summary = globalSummary(profile: profile, commanders: commanders)
        return .init(
            stageStarsEarned: stars.earned,
            stageStarsMax: stars.max,
            military: summary.military,
            nobility: summary.nobility,
            continents: [
                "europe": continent(profile: profile, key: "europe"),
                "america": continent(profile: profile, key: "america"),
                "asia": continent(profile: profile, key: "asia")
            ],
            generalIDs: ownedGeneralIDs(profile: profile)
        )
    }
}

File: SOURCE_LEAN/EW4_Web_Port_v0.61/native_achievement_core.js
'use strict';
(function(root,factory){
  const general=(typeof module==='object'&&module.exports)?require('./native_general_core.js'):root.EW4NativeGeneral;
  const api=factory(general);
  if(typeof module==='object'&&module.exports)module.exports=api;
  root.EW4NativeAchievement=api;
})(typeof globalThis!=='undefined'?globalThis:this,function(General){
  const MILITARY_THRESHOLDS=Object.freeze([...(General?.MILITARY_THRESHOLDS||[500,800,1200,1900,3000,4800,7500,12000,19000,30000,48000,76000,120000,200000])]);
  const NOBILITY_THRESHOLDS=Object.freeze([...(General?.NOBILITY_THRESHOLDS||[100,200,300,450,675,1000,1500,2250,3375])]);
  const STAGE_COUNTS=Object.freeze([18,17,15,13,11,10]);
  const MAX_STAGE_STARS=STAGE_COUNTS.reduce((a,n)=>a+n*5,0); // native: 84 stages * 5 = 420
  const obj=v=>v&&typeof v==='object'&&!Array.isArray(v)?v:{};
  const clamp=(v,lo,hi)=>Math.max(lo,Math.min(hi,Math.trunc(+v||0)));
  const nonneg=v=>Math.max(0,Math.trunc(+v||0));
  const own=(o,k)=>Object.prototype.hasOwnProperty.call(o||{},k);

  function ownedGeneralIds(save,maxId=208){
    const seen=new Set(),out=[];
    for(const raw of Array.isArray(save?.owned)?save.owned:[]){
      const id=Math.trunc(+raw||0);if(id<1||id>maxId||seen.has(id))continue;seen.add(id);out.push(id);
    }
    return out;
  }
  function commanderLookup(commanders,id){
    if(!commanders)return null;
    if(Array.isArray(commanders))return commanders.find(c=>+c?.id===+id)||null;
    return commanders[String(id)]||commanders[id]||null;
  }
  function effectiveGeneralState(save,commanders,id){
    const c=commanderLookup(commanders,id)||{};
    const ranks=obj(save?.rank),nobs=obj(save?.nobility),rprog=obj(save?.rankProgress),nprog=obj(save?.nobilityProgress);
    return{
      id,
      rank:clamp(own(ranks,id)?ranks[id]:(c.rank??0),0,MILITARY_THRESHOLDS.length),
      rankProgress:nonneg(rprog[id]),
      nobility:clamp(own(nobs,id)?nobs[id]:(c.nobilityrank??c.nobility??0),0,NOBILITY_THRESHOLDS.length),
      nobilityProgress:nonneg(nprog[id])
    };
  }
  function completedRaw(level,progress,thresholds,base){
    level=clamp(level,0,thresholds.length);let total=base+nonneg(progress);for(let i=0;i<level;i++)total+=thresholds[i];return total;
  }
  function rawMilitary(state){return completedRaw(state?.rank,state?.rankProgress,MILITARY_THRESHOLDS,300)}
  function rawNobility(state){return completedRaw(state?.nobility,state?.nobilityProgress,NOBILITY_THRESHOLDS,60)}
  function nativeFloatThresholdNext(threshold,multiplier){
    return Math.trunc(Math.fround(Math.fround(threshold)*Math.fround(multiplier)));
  }
  function aggregateLevel(total,{cutoff,first,multiplier}){
    total=nonneg(total);if(total<=cutoff)return 1;
    let level=2,threshold=first;
    while(level<99&&total>=threshold){threshold=nativeFloatThresholdNext(threshold,multiplier);level++}
    return level;
  }
  function globalSummary(save,commanders){
    // Native EW4 iterates the original 12 HQ slots. This port deliberately removes
    // the general-cap mod, so the same native per-general/raw/global formulas are
    // extended across every owned HQ general instead of silently ignoring slot 13+.
    const states=ownedGeneralIds(save).map(id=>effectiveGeneralState(save,commanders,id));
    const militaryRaw=states.reduce((a,s)=>a+rawMilitary(s),0),nobilityRaw=states.reduce((a,s)=>a+rawNobility(s),0);
    return{
      military:{level:aggregateLevel(militaryRaw,{cutoff:99,first:121,multiplier:1.214}),score:Math.floor(militaryRaw/10),raw:militaryRaw},
      nobility:{level:aggregateLevel(nobilityRaw,{cutoff:49,first:56,multiplier:1.125}),score:nobilityRaw*4,raw:nobilityRaw},
      countedGenerals:states.length
    };
  }
  function legacyBestRating(save,file){
    const best=obj(save?.campaignBestRating),progress=obj(save?.campaignProgress);
    if(own(best,file))return clamp(best[file],0,5);
    const p=clamp(progress[file],0,2);return p>=2?5:p>=1?1:0;
  }
  function campaignStageStars(save){
    let earned=0;
    for(let z=1;z<=STAGE_COUNTS.length;z++)for(let s=1;s<=STAGE_COUNTS[z-1];s++)earned+=legacyBestRating(save,`campaign${z}_${String(s).padStart(2,'0')}.btl`);
    return{earned:clamp(earned,0,MAX_STAGE_STARS),max:MAX_STAGE_STARS};
  }
  function continentStoredValue(save,key){
    const r=obj(save?.achievementConquests)[key];
    if(Number.isFinite(+r))return clamp(r,0,1000);
    if(r&&typeof r==='object'){
      if(Number.isFinite(+r.value))return clamp(r.value,0,1000);
      // Compatibility with the conservative P27 shape; only migrate its rendered
      // digits, never the invented separate `year` field.
      if(Array.isArray(r.rule)&&r.rule.length){let n=0;for(const d of r.rule.slice(0,3))n=n*10+clamp(d,0,9);return clamp(n,0,1000)}
    }
    return 0;
  }
  function ruleDigits(value){
    const n=clamp(value,0,999);return String(n).split('').map(x=>+x);
  }
  function continent(save,key){
    const stored=continentStoredValue(save,key),display=Math.min(999,stored);
    return{value:stored,display,digits:ruleDigits(display)};
  }
  function recordConquestValue(save,key,value){
    if(!['europe','america','asia'].includes(key))throw new Error('invalid continent');
    const next=Object.assign({},save),records=Object.assign({},obj(save?.achievementConquests)),old=continentStoredValue(save,key),v=clamp(value,0,1000);
    if(v>old)records[key]={value:v};next.achievementConquests=records;return{save:next,changed:v>old,value:Math.max(old,v)};
  }
  function viewModel(save,commanders){
    const summary=globalSummary(save,commanders);
    return{
      stageStars:campaignStageStars(save),
      walletStars:clamp(save?.campaignStars,0,999),
      military:summary.military,
      nobility:summary.nobility,
      countedGenerals:summary.countedGenerals,
      continents:{europe:continent(save,'europe'),america:continent(save,'america'),asia:continent(save,'asia')},
      generalIds:ownedGeneralIds(save)
    };
  }
  return Object.freeze({MILITARY_THRESHOLDS,NOBILITY_THRESHOLDS,STAGE_COUNTS,MAX_STAGE_STARS,ownedGeneralIds,effectiveGeneralState,rawMilitary,rawNobility,aggregateLevel,globalSummary,campaignStageStars,continentStoredValue,ruleDigits,continent,recordConquestValue,viewModel});
});
